xline/src/server/xline_server.rs

```diff
@@ -285,124 +238,133 @@ where
+    )> {
+        let (header_gen, id_gen, auth_revision_gen) = Self::construct_generator(&self.cluster_info);
+        let lease_collection = Self::construct_lease_collection(
+            self.curp_cfg.heartbeat_interval,
+            self.curp_cfg.candidate_timeout_ticks,
+        );
+
+        let (kv_storage, lease_storage, auth_storage, watcher) =
+            Self::construct_underlying_storages(
+                self.is_leader,
+                &persistent,
+                lease_collection,
+                &header_gen,
+                &auth_revision_gen,
+                key_pair,
+            )?;
+        let client = Arc::new(
+            Client::<Command>::new(self.cluster_info.get_all_members(), self.client_timeout).await,
+        );
+        let index_barrier = Arc::new(IndexBarrier::new());
+        let id_barrier = Arc::new(IdBarrier::new());
+
+        let state = State::new(Arc::clone(&lease_storage));
+        let ce = CommandExecutor::new(
+            Arc::clone(&kv_storage),
+            Arc::clone(&auth_storage),
+            Arc::clone(&lease_storage),
+            Arc::clone(&persistent),
+            Arc::clone(&index_barrier),
+            Arc::clone(&id_barrier),
+            header_gen.revision_arc(),
+            Arc::clone(&auth_revision_gen),
+        );
+
+        Ok((
+            KvServer::new(
+                Arc::clone(&kv_storage),
+                Arc::clone(&auth_storage),
+                Arc::clone(&index_barrier),
+                Arc::clone(&id_barrier),
+                self.range_retry_timeout,
+                Arc::clone(&client),
+                self.cluster_info.get_self_id().clone(),
+            ),
+            LockServer::new(
+                Arc::clone(&client),
+                Arc::clone(&id_gen),
+                self.cluster_info.get_self_id().clone(),
+                self.cluster_info.get_self_address().to_owned(),
+            ),
+            LeaseServer::new(
+                Arc::clone(&lease_storage),
+                Arc::clone(&auth_storage),
+                Arc::clone(&client),
+                Arc::clone(&id_gen),
+                Arc::clone(&self.cluster_info),
+            ),
+            AuthServer::new(
+                Arc::clone(&auth_storage),
+                Arc::clone(&client),
+                self.cluster_info.get_self_id().clone(),
+            ),
+            WatchServer::new(Arc::clone(&watcher), Arc::clone(&header_gen)),
+            MaintenanceServer::new(Arc::clone(&persistent), Arc::clone(&header_gen)),
+            Self::construct_curp_server(
+                Arc::clone(&self.cluster_info),
+                self.is_leader,
+                ce,
+                state,
+                &self.storage_cfg,
+                Arc::clone(&self.curp_cfg),
+            )
+            .await,
+        ))
+    }
+
+    /// Construct a new `CurpServer`
+    async fn construct_curp_server<S: StorageApi, CE: CurpCommandExecutor<Command> + 'static>(
```

Comment:
Constuctor could take a &self to reduce the number of parameters. Same for `construct_underlying_storages`.

Reply:
The most of `construct_*` functions are not depend too much on the `XlineServer`. Therefore, it may look wired to make `construct_curp_server` become one of the `XlineServer` methods while the others are still functions.

Reply:
Makes sense.